pkg/store/gcs.go

```diff
@@ -182,151 +200,59 @@ func (s *GCSStore) Close() (err error) {
+				if err != nil {
+					level.Warn(s.logger).Log("msg", "loading block failed", "id", id, "err", err)
+					// Wipe the directory so we can cleanly try again later.
+					os.RemoveAll(dir)
+					continue
+				}
+				s.setBlock(id, b)
 			}
 			wg.Done()
-			<-workc
 		}()
 	}
-	wg.Wait()
 
-	return nil
-}
-
-// downloadBlock downloads the index and meta.json file for the given block ID and opens a reader
-// against the block.
-func (s *GCSStore) downloadBlock(ctx context.Context, id ulid.ULID) error {
-	bdir := filepath.Join(s.dir, id.String())
-	tmpdir := bdir + ".tmp"
-
-	if err := os.MkdirAll(tmpdir, 0777); err != nil {
-		return errors.Wrap(err, "create temp dir")
-	}
-
-	for _, fn := range []string{
-		"index",
-		"meta.json",
-	} {
-		obj := s.bucket.Object(path.Join(id.String(), fn))
-
-		f, err := os.Create(filepath.Join(tmpdir, fn))
-		if err != nil {
-			return errors.Wrap(err, "create local index copy")
-		}
-
-		s.metrics.gcsOperations.WithLabelValues(gcsOperationObjectGet).Inc()
-		r, err := obj.NewReader(ctx)
-		if err != nil {
-			return errors.Wrap(err, "create index object reader")
-		}
-		_, copyErr := io.Copy(f, r)
-
-		if err := f.Close(); err != nil {
-			level.Warn(s.logger).Log("msg", "close file", "err", err)
-		}
-		if err := r.Close(); err != nil {
-			level.Warn(s.logger).Log("msg", "close object reader", "err", err)
-		}
-		if copyErr != nil {
-			if err := os.RemoveAll(tmpdir); err != nil {
-				level.Warn(s.logger).Log("msg", "cleanup temp dir after failure", "err", err)
-			}
-			return errors.Wrap(copyErr, "copy index file to disk")
+Loop:
+	for {
+		oi, err := objs.Next()
+		if err == iterator.Done {
+			break
+		} else if err != nil {
+			close(blockc)
+			return err
 		}
-	}
-
-	if err := renameFile(tmpdir, bdir); err != nil {
-		return errors.Wrap(err, "rename block directory")
-	}
-	return nil
-}
-
-// loadBlocks ensures that all blocks in the data directory are loaded into memory.
-// Additionally, it saves metadata about the oldest and newest block.
-func (s *GCSStore) loadBlocks() error {
-	oldestBlockMinTime := s.oldestBlockMinTime
-	youngestBlockMaxTime := s.youngestBlockMaxTime
-
-	fns, err := fileutil.ReadDir(s.dir)
-	if err != nil {
-		return errors.Wrap(err, "read dir")
-	}
-	for _, fn := range fns {
-		id, err := ulid.Parse(fn)
+		id, err := ulid.Parse(oi.Prefix[:len(oi.Prefix)-1])
 		if err != nil {
 			continue
 		}
 		if b := s.getBlock(id); b != nil {
 			continue
 		}
-		b, err := s.loadFromDisk(id)
-		if err != nil {
-			level.Warn(s.logger).Log("msg", "loading block failed", "err", err)
-			continue
-		}
-		s.setBlock(id, b)
-
-		if oldestBlockMinTime > b.meta.MinTime || oldestBlockMinTime == math.MaxInt64 {
-			oldestBlockMinTime = b.meta.MinTime
-		}
-
-		if youngestBlockMaxTime < b.meta.MaxTime || youngestBlockMaxTime == math.MaxInt64 {
-			youngestBlockMaxTime = b.meta.MaxTime
+		select {
+		case <-ctx.Done():
+			break Loop
```

Comment:
instead of not often used loop labels we can just do
```
close(blockc)
wg.Wait()
return nil
```

But it is up to you (: